dynamic_programming/unique_paths.cpp

```diff
@@ -0,0 +1,132 @@
+/**
+ * @file
+ * @brief Implementation of Unique Paths (LeetCode 62) using Dynamic Programming.
+ * @details
+ * A robot is located at the top-left corner of an m × n grid.
+ * The robot can move either down or right at any point in time.
+ * This program computes the total number of unique paths to reach
+ * the bottom-right corner.
+ *
+ * Note: This is **Unique Paths I** (no obstacles). Obstacles are handled in
+ * the separate problem **Unique Paths II**.
+ *
+ * Approaches:
+ * - **Memoization (Top-Down)**: Recursively explores solutions while
+ *   storing intermediate results in a cache (`memoization_table`) to avoid redundant
+ *   computation. Typically more intuitive and easy to write, but
+ *   relies on recursion and has associated call-stack overhead.
+ *
+ * - **Tabulation (Bottom-Up)**: Iteratively builds the solution using
+ *   a DP table, starting from the base cases and filling up the table.
+ *   Offers consistent performance without recursion overhead and
+ *   is generally more space-efficient when optimized.
+ *
+ * @see https://leetcode.com/problems/unique-paths/
+ */
+
+#include <iostream>
+#include <vector>
+#include <cassert>
+
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming algorithms
+ */
+namespace dynamic_programming {
+
+/**
+ * @class UniquePathsSolver
+ * @brief Solves the Unique Paths problem using both memoization and tabulation.
+ */
+class UniquePathsSolver {
+   private:
+    std::vector<std::vector<int>> memoization_table;  ///< Memoization table to cache intermediate results
+    std::size_t m;  ///< Number of rows in the grid
+    std::size_t n;  ///< Number of columns in the grid
+
+    /**
+     * @brief Bottom-up Tabulation solution.
+     * @return int Number of unique paths from (0, 0) to (m-1, n-1)
+     */
+    int solveTabulation() {
+        std::vector<std::vector<int>> table(m, std::vector<int>(n, 0));
+
+        for (std::size_t i = 0; i < m; i++) table[i][n - 1] = 1;  ///< last column
+        for (std::size_t j = 0; j < n; j++) table[m - 1][j] = 1;  ///< last row
+
+        for (int i = static_cast<int>(m) - 2; i >= 0; i--) {
+            for (int j = static_cast<int>(n) - 2; j >= 0; j--) {
```

Comment:
Instead of the `static_cast` present here, You could change the type of the iterator variables, `i` and `j` to `size_t` as well.